goto.1
.\"
.\" Copyright (c) 2004-2011
.\"	Jeffrey Allen Neitzel <jan (at) v6shell (dot) org>.
.\"	All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY JEFFREY ALLEN NEITZEL ``AS IS'', AND ANY
.\" EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
.\" WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
.\" DISCLAIMED.  IN NO EVENT SHALL JEFFREY ALLEN NEITZEL BE LIABLE FOR ANY
.\" DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
.\" LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
.\" AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
.\" OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
.\" USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\"	@(#)$Id$
.\"
.\"	Derived from: Sixth Edition UNIX /usr/man/man1/goto.1
.\"
.\" Copyright (C) Caldera International Inc.  2001-2002.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code and documentation must retain the above
.\"    copyright notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. All advertising materials mentioning features or use of this software
.\"    must display the following acknowledgement:
.\"      This product includes software developed or owned by Caldera
.\"      International, Inc.
.\" 4. Neither the name of Caldera International, Inc. nor the names of other
.\"    contributors may be used to endorse or promote products derived from
.\"    this software without specific prior written permission.
.\"
.\" USE OF THE SOFTWARE PROVIDED FOR UNDER THIS LICENSE BY CALDERA
.\" INTERNATIONAL, INC. AND CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR
.\" IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
.\" OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
.\" IN NO EVENT SHALL CALDERA INTERNATIONAL, INC. BE LIABLE FOR ANY DIRECT,
.\" INDIRECT INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
.\" (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
.\" SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
.\" STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
.\" IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
.\" POSSIBILITY OF SUCH DAMAGE.
.\"
.TH GOTO 1 "@OSH_DATE@" "@OSH_VERSION@" "General Commands"
.SH NAME
goto \- transfer command
.SH SYNOPSIS
.BI
.B goto
\fIlabel\fR [\fI...\fR]
.SH DESCRIPTION
.I Goto
is allowed only when the shell is taking
commands from a seekable command file (see
.IR lseek (2)).
.PP
The entire command file is searched for a line beginning
with a \fB:\fR as the first non-blank character,
followed by one or more blanks,
and then the
.IR label .
If such a line is found,
.I goto
repositions the command-file offset to
the line after the label and exits.
This effectively causes the shell to transfer
to the labelled line.
.PP
\fB:\fR is a special command which causes
the shell to do nothing.
.SH "EXIT STATUS"
The
.I goto
command exits with one of the following values:
.TP
0
The label was found.
.TP
1
The label was not found.
.TP
2
An error was detected.
.SH "SEE ALSO"
if(1),
osh(1),
sh6(1)
.PP
Osh home page:
http://v6shell.org/
.SH COMPATIBILITY
The
.I goto
command from Sixth Edition UNIX
requires \fB:\fR to be the first character
on the labelled line.
Otherwise,
the label cannot be found.
.PP
This port differs in that \fB:\fR may be preceded by blanks.
This allows labelled lines to be indented so that complex
command files might be easier to read and understand.
.SH HISTORY
A
.I goto
command
appeared as
.I /bin/goto
in Third Edition UNIX.
.SH LICENSE
See either the LICENSE file which is distributed with
.I osh
or
http://v6shell.org/license/
for full details.
.SH COPYRIGHT
.nf
Copyright (c) 2004-2011
	Jeffrey Allen Neitzel.  All rights reserved.

Copyright (c) 2001-2002
	Caldera International Inc.  All rights reserved.
.fi


CHANGES
Changes that improve compatibility w/ the Sixth Edition (V6) UNIX
Thompson shell are marked w/ a `C:' in the details below.

-------------------------------------------------------------------------------
[osh-current]: Friday, 2010-05-14:
osh.1:
	* Added [--] to fd2 usage synopsis in "Special commands" subsection.

	* s/Mac/Mac OS X/

-------------------------------------------------------------------------------
[osh-20100430]:
osh.c:
	* Added a new, osh history-file-support feature to save the user's
	  command-line history to the user $h/.osh.history file.  Notice
	  that the history file must already exist as a writable, regular
	  file (or a link to a regular file) when the shell is invoked to
	  save the user's command-line history.  Otherwise, it won't do so.

	  An interactive shell reads each command line from its terminal and
	  writes the words of each one to the history file as a history entry
	  after performing parameter substitution and word splitting.

osh.1:
	* Documented the new, osh history-file-support feature in the new
	  "History file (+)" subsection.

-------------------------------------------------------------------------------
[osh-20100228]:
mkconfig:
	* Added a search comparison that uses which(1) to find login & newgrp
	  pathname(s).  This allows for those cases where a sysadmin may use
	  locally modified versions of these utilities to deal w/ uniquely
	  local needs.  Notice that this is in addition to how login & newgrp
	  pathname(s) were searched for before now.

osh.c:
	* Changed osh globbing so it's handled in main osh process.  This is
	  for user convenience to allow (for example):

		% mkdir some_really_long_directory_name... ; cd some*...

	  ...
	  and globbing for other special built-in commands as well.

-------------------------------------------------------------------------------
[osh-20091218]:
	* Incorporated osh-20091127-p1.patch changes (see p1: below).

INSTALL, Makefile:
	* Made "install-doc" and "install-exp" targets optional so that
	  they're no longer triggered by default "install" target(s).

mkconfig:
	* p1: Defined CONFIG_SUNOS on OpenSolaris (uname -s == SunOS).

goto.c:
	* p1: Fixed problem w/ external goto command on OpenSolaris where
	      file offset set via lseek(2) isn't shared w/ parent process,
	      causing unexpected behavior.

examples/*osh*:
	* Made changes that allow users to use osh to initialize/invoke sh6
	  by doing `[exec] osh [-i | -l] sh6'. This is most convenient way
	  to perform startup actions for sh6 without adding undesirable
	  functionality to sh6 itself.

examples/dot.osh.logout:
	* Fixed `rm -r $d' error on OpenSolaris.

-------------------------------------------------------------------------------
[osh-20091127]:
err.c:
	* Changed setmyname(argv[0]) so that it only ditches an initial
	  "-" character when a binary name is "-" plus at least 1 of any
	  character before end-of-string.  For example:

		% pwd ; ln -s osh - ; ln -s osh -o ; ln -s osh -os ; \
			ln -s osh -osh ; echo -* osh
		/tmp/v6shell
		- -o -os -osh osh

		% which - -o -os -osh osh
		/tmp/v6shell/-
		/tmp/v6shell/-o
		/tmp/v6shell/-os
		/tmp/v6shell/-osh
		/tmp/v6shell/osh

		% - file arg1 ; -o file arg1 ; -os file arg1 ; \
			-osh file arg1 ; osh file arg1
		-: file: cannot open
		o: file: cannot open
		os: file: cannot open
		osh: file: cannot open
		osh: file: cannot open

-------------------------------------------------------------------------------
[osh-20091122]:
defs.h:
	* Defined SBUFMM(m), small buffer max multiplier, for calculating
	  small buffer sizes (e.g., DOLMAX, FMTMAX, LABELMAX).

err.h:
	* Changed ERR_ALTOOLONG to ERR_E2BIG, and added FMT4S.

pexec.[hc]:
	* Changed err_pexec() function prototype.

	* Changed it to print ERR_E2BIG diagnostic when errno == E2BIG.

	* Changed exit status to 126 for such "cannot execute" conditions.

*.c:
	* Changed all err_pexec() calls to match new function prototype
	  where appropriate.

osh.c:
	* Changed main child process execute function from pexec()
	  to err_pexec() for external commands.

	* Added getmyname() program name prefix (e.g., "osh: ")
	  to osh diagnostics.

v.c:
	* Added "The Regents of the University of California" copyright notice
	  for NetBSD and OpenBSD code included in the osh project.

[fgio][dfos]*.1:
	* Added COPYRIGHT section headings, and added copyright notices.

-------------------------------------------------------------------------------
[osh-20091029]:
	* Remove rcsid.h file; stop adding rcsid strings for all
	  sources to compiled binaries.

	* Re-add example rc (init and logout) files as examples/* .

	* Add $h/.project file as PROJECT for quick project info.

Makefile:
	* Add default DOCDIR as $(PREFIX)/share/doc/$(OSH_VERSION);
	  use it for [ACILNPR]* package documentation and examples/* .

	* Change default MANDIR to $(PREFIX)/share/man/man1 .

v.c:
	* Include OSH_VERSION from Makefile and project copyright
	  notices in compiled binaries.

fd2.c, if.c, pexec.[hc], util.c:
	* Add err_pexec() error-handling wrapper around pexec() for
	  shell utilities that execute external commands.

osh.c:
	* Wrap WCOREDUMP(s) w/ #ifdef WCOREDUMP ... #endif .

osh.1:
	* Change example in "I/O redirection" subsection that could
	  confuse new users.

-------------------------------------------------------------------------------
[osh-20090527]:
	* Moved inclusion of required kernel include files (e.g., <sys/*.h>)
	  and /usr/include files from *.c to "defs.h" where appropriate.

	* Moved files:	Moved manual pages from *.1.in to *.1 .

	* New files:	err.h, err.c  (see err.[hc] description below)

Makefile:
	* Changed "-O2" to "-Os" in CFLAGS to optimize for size.

	* Changed how the manual pages are generated and installed.

README:
	* Added Debian 5.0.1 (lenny), NetBSD 5.0, OpenBSD 4.5, and
	  OpenSolaris 2008.11 to the reported system list; the osh
	  package behaves as expected on each system.

err.[hc]:

goto.c
/*
 * goto.c - a port of the Sixth Edition (V6) UNIX transfer command
 */
/*-
 * Copyright (c) 2004-2011
 *	Jeffrey Allen Neitzel <jan (at) v6shell (dot) org>.
 *	All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY JEFFREY ALLEN NEITZEL ``AS IS'', AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
 * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED.  IN NO EVENT SHALL JEFFREY ALLEN NEITZEL BE LIABLE FOR ANY
 * DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
 * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED
 * AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
 * OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
 * USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 *	@(#)$Id$
 */
/*
 *	Derived from: Sixth Edition UNIX /usr/source/s1/goto.c
 */
/*-
 * Copyright (C) Caldera International Inc.  2001-2002.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code and documentation must retain the above
 *    copyright notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. All advertising materials mentioning features or use of this software
 *    must display the following acknowledgement:
 *      This product includes software developed or owned by Caldera
 *      International, Inc.
 * 4. Neither the name of Caldera International, Inc. nor the names of other
 *    contributors may be used to endorse or promote products derived from
 *    this software without specific prior written permission.
 *
 * USE OF THE SOFTWARE PROVIDED FOR UNDER THIS LICENSE BY CALDERA
 * INTERNATIONAL, INC. AND CONTRIBUTORS ``AS IS'' AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 * IN NO EVENT SHALL CALDERA INTERNATIONAL, INC. BE LIABLE FOR ANY DIRECT,
 * INDIRECT INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
 * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
 * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
 * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
 * POSSIBILITY OF SUCH DAMAGE.
 */

#include "defs.h"
#include "err.h"

#ifndef	CONFIG_SUNOS
static	off_t	offset;
#endif

static	bool	getlabel(/*@out@*/ char *, int, size_t);
static	int	xgetc(void);

/*
 * NAME
 *	goto - transfer command
 *
 * SYNOPSIS
 *	goto label [...]
 *
 * DESCRIPTION
 *	See the goto(1) manual page for full details.
 */
int
main(int argc, char **argv)
{
	size_t siz;
	char label[LABELMAX];

	setmyerrexit(&ut_errexit);
	setmyname(argv[0]);
	setmypid(getpid());

	if (argc < 2 || *argv[1] == EOS || isatty(FD0) != 0)
		err(FC_ERR, FMT2S, getmyname(), ERR_GENERIC);
	if ((siz = strlen(argv[1]) + 1) > sizeof(label))
		err(FC_ERR, FMT3S, getmyname(), argv[1], ERR_LABTOOLONG);
	if (lseek(FD0, (off_t)0, SEEK_SET) == -1)
		err(FC_ERR, FMT2S, getmyname(), ERR_SEEK);

	while (getlabel(label, *argv[1] & 0377, siz))
		if (strcmp(label, argv[1]) == 0) {
#ifndef	CONFIG_SUNOS
			(void)lseek(FD0, offset, SEEK_SET);
#endif
			return SH_TRUE;
		}

	fd_print(FD2, FMT3S, getmyname(), argv[1], ERR_LABNOTFOUND);
	return SH_FALSE;
}

/*
 * Search for the first occurrence of a possible label with both
 * the same first character (fc) and the same length (siz - 1)
 * as argv[1], and copy this possible label to buf.
 * Return true  (1) if possible label found.
 * Return false (0) at end-of-file.
 */
static bool
getlabel(char *buf, int fc, size_t siz)
{
	int c;
	char *b;

	while ((c = xgetc()) != EOF) {
		/* `:' may be preceded by blanks. */
		while (c == SPACE || c == TAB)
			c = xgetc();
		if (c != COLON) {
			while (c != EOL && c != EOF)
				c = xgetc();
			continue;
		}

		/* Prepare for possible label. */
		while ((c = xgetc()) == SPACE || c == TAB)
			;	/* nothing   */
		if (c != fc)	/* not label */
			continue;

		/*
		 * Try to copy possible label (first word only)
		 * to buf, ignoring it if it becomes too long.
		 */
		b = buf;
		do {
			if (c == EOL || c == SPACE || c == TAB || c == EOF) {
				*b = EOS;
				break;
			}
			*b = c;
			c = xgetc();
		} while (++b < &buf[siz]);

		/* Ignore any remaining characters on labelled line. */
		while (c != EOL && c != EOF)
			c = xgetc();
		if (c == EOF)
			break;

		if ((size_t)(b - buf) != siz - 1)	/* not label */
			continue;
		return true;
	}

	*buf = EOS;
	return false;
}

/*
 * If not at end-of-file, return the next character from the standard
 * input as an unsigned char converted to an int while incrementing
 * the global offset.  Otherwise, return EOF at end-of-file.
 */
static int
xgetc(void)
{
#ifndef	CONFIG_SUNOS
	int nc;

	offset++;
	nc = getchar();
	return (nc != EOF) ? nc & 0377 : EOF;
#else
	unsigned char nc;

	return (read(FD0, &nc, (size_t)1) == 1) ? nc : EOF;
#endif
}
